rfcs/0052-dynamic-ids.md

```diff
@@ -0,0 +1,112 @@
+---
+feature: dynamic-ids
+start-date: 2019-09-05
+author: Silvan Mosberger
+co-authors: (find a buddy later to help our with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: https://github.com/NixOS/nixpkgs/pull/65698
+---
+
+# Summary
+[summary]: #summary
+
+A lot of NixOS modules are [assigning static uids/gids](https://github.com/NixOS/nixpkgs/blob/044cc701c23ede96355eb1dc997985b4dfac0372/nixos/modules/misc/ids.nix#L36) to their users. This has resulted in less than 90 static ids left in the reserved range from 0 to 400.
+
+This RFC deprecates the practice of doing that and instead suggests to
+- If applicable use systemd's `DynamicUser`
+- Otherwise let NixOS assign dynamic persistent ids, which happens automatically when `users.users.<name?>.uid`/`users.groups.<name?>.gid` is not set. For users, `users.users.<name?>.isSystemUser` should be set so that only uids under 1000 are used by NixOS services.
+
+Only in special circumstances are static ids allowed anymore.
+
+# Motivation
+[motivation]: #motivation
+
+We are running out of static ids in the reserved range from 0 to 400. If services continue to reserve ids for themselves this will run out eventually.
+
+In addition, a central list of ids is annoying to maintain and leads to merge conflicts.
+
+# Detailed design
+[design]: #detailed-design
+
+## Documentation updates
+
+New best practices for declaring users are to be documented. This includes sections like the following
+
+### Using `DynamicUser` for declaring users
+
+If the service is fit for `DynamicUser`, this is the preferred solution, as it doesn't require any persistent ids. However this only works well for services whose state is self-contained. As soon as the services data is needed somewhere else, this approach can fall flat. `DynamicUser` also enables a bunch of options restricting the service (see `man systemd.exec` for details), some of which might need to be disabled in order for it to work.
+
+An example:
+
+```nix
+{
+  systemd.services.myservice.serviceConfig.DynamicUser = true;
+}
+```
+
+### Using ids dynamically allocated by NixOS
+
+
+By not setting `users.users.<name?>.uid`/`users.groups.<name?>.gid`, the ids will be dynamically allocated by NixOS. For users, `users.users.<name?>.isSystemUser = true` should be set as well such that only ids below 1000 are to be used. These ids are persistent over the lifetime of a NixOS system, even when services are disabled and enabled again. The generated mapping from names to ids is stored in `/var/lib/nixos/uid-map`/`/var/lib/nixos/gid-map`, so if this directory is backed up, the mappings will persist too when restoring.
```

Comment:
It would be great if we can defer this if possible and only use dynamically allocated uids if absolute necessary. Users may also allocate on their own static uids/gids for system users to avoid the issue when migrating data between machines.

Reply:
Just to throw in an idea:
If there would be a mechanism that would allow users to put ad-hoc allocated uid/gids into the nixos configuration, than this information could be easily backed and put into version control.

Reply:
That's always possible by assigning `users.users.<name?>.uid` in their config. I guess I haven't mentioned this, but there's nothing wrong with the end user using static ids.

However I think it would be good if we could get away from static ids completely and not object anybody to decide on these arbitrary numbers. When things like `StateDirectory` are used throughout, non-static ids won't be a problem even when migrating data.